README.md
Gmail-to-Trello Extension
=========================

Work on this repo has been abandoned for a large name change we had to undergo due to us being unduely affected by Chrome Web Store hacking. Please join our open source efforts here: http://g2t.pub/github and our community here: http://g2t.support.


chrome/CHANGES.md
A free tool that provides an extra button on Gmail UI to add gmail text and attachments to a Trello card

=== 2.8.0.005@2020-04-19 ===
 • Had to publish as new Extension due to old extension updates being disallowed.
 • Change "Arrows" to a pop-up for "Add below:" and "Add to:"
 • Try AspectRatio on for more reasonable resize experience [nope, doesn't work for what we'd want]
 
=== 2.8.0.003@2020-03-04 ===
 • Now uses background script for retrieval and upload of all attachments/images/embedded images, should resolve CORS/CORB new higher level of Chrome Security.
 • Standardize on selectors for embedded titles
 • Reduce scopes of domains in manifest
 • Update descriptive text in manifest
 • Update to jQuery 3.4.1.slim
 • Add option/shift-click to "Attach" or "Images" will select-all/deselect-all based on first checkbox's state
 • Remove in-app purchases
 • Output jQuery object when body can't be found

=== 2.7.2.32@2020-02-15...2.7.0.0 ===
 • Try some changes to grow boxes (unsuccessfully)
 • Update file naming upload code to gleen filename from aria-label text
 • Gmail UI class names changed again, in gmailView we now use a single viewport: '.aia:first'. It was '.aeJ:first'. We no longer explicitly try to detect splitlayout, which may or may not work. Will need to test with folks.
 • Initial error message when retrieving attachment content fails with 0 length. Probably CORS/CORB new Chrome security model, need to retrieve data from background script. Fix in progress but will take a while to refactor uploading code.
 • Deeper highlighting for Labels and Assign, use gradient to indicate selected
 • Remove trash can from GtT pull down
 • Move popup location of GtT pull down to under icon
 • Use new === of Trello Client.js
 • Make sure UI, Sign-out works (enough) when Trello Authenticate fails.
 • Look for old legacy classname m{UNIQUEID} only if other emailId tags aren't present
 • Iterate differently
 • Use data-thread-id if present
 • Get Legacy Gmail Thread ID for backlink on new Gmail
 • Find new Gmail views
 • Make getManifest === check more robust error-handling-wise
 • Fix calculation for pop-up when button position has changed
 • New HD toolbar icon & sketch version
 • Attachment image previews are now scale-to-fit
 • Attachment file and image sections are hidden when there is no attachment
 • [dev] Change form markups from dl/dt/dd to regular divs
 • Fix bug on the new Gmail that cause toolbar buttons failed to work properly
 • Remove overlapping intervals, thanks to Travis Hardman.
 • Up debug log
 • Silence logging of button positioner [still need to fix]
 • Gmails coming up blank content in GtT - Change .adP:first to use more deterministic tag
 • Check for ASL or ASF div for Refresh icon.
 • Only call detach if more than one button/popup
 • Use '0' for unknown === in === check
 • More debugging logs for adding button / popup
 • Only show === update if previous === > 0
 • Remove toolBarHolder
 • Debug Multiple Inboxes - more to do
 • Circular log wasn't - hogged memory
 • inline image now should upload to Trello correctly
 • Restrict log from 1000 to 100 lines.
 • Call browsingData from background to clear extension data.
 • Show message to reload when === changes.
 • Fix $button[0] -> $button.first().
 • Fix missing views/options.html.
 • detectToolbar return true if detected.
 • .detach button and popup and then only append one.
 • Make sign out an explicit button on the page.
 • Fix resize via jQuery UI, needed clearfix at popup level for jQuery UI added elements
 • Call pre-init after button disappear
 • Semi-final jQuery UI resizing (not quite right for normal state, works okay for list state)
 • Show === number in options panel (prep for noticing === change and prompting to reload)
 • Remove G-Ni from GtT icon
 • Add timer to check every 2 seconds for GtT button showing
 • Use gh='mtb' to find toolbar
 • Add G-Ni to GtT icon
 • Remove :first from G-atb
 • Additional debugging code in toolbar and labels code
 • Report of GtT button not appearing when Streak and RightInbox, some tweaks to button positioning code to hopefully circumvent problem.
 • Add info message to options screen
 • Initial cut at button for chrome.browsingData.remove
 • Update manifest with 64 and 128 icons
 • Icon not showing up in toolbar due to other extension icons in toolbar
 • Update gtt_log to have timestamp
 • Move email search hyperlinks to top of content
 • Parse "29. Mai 2017 um 15:18" correctly
 • (Until I have a better idea, have to decode the dateTime by hand and do some month comparisons)
 • And then after doing that ton of work and realizing this will be horrific to maintain, I ripped it all out
 • Check for 401 more leniently, add target to error
 • Lists without organizations were being filtered out! Fixed.
 • Change Features/Bugs to 'Help'
 • Created 'Report' feature, which will put latest error and last 1000 log items into card to post to GtT Trello board
 • 400 invalid id on attachment upload: Use pos === 'at' to indicate path to attach
 • Track mouseUp and mouseDown in same external-to-window container
 • Make positioning logic more robust for Upload to combat Trello POST pos 404 error
 • Click outside window closes window
 • Focus outside window closes window
 • Error in attachments processing fixed to produce correct filename
 • Install keyboard trap to Show Popup, Remove keyboard trap on Hide Popup
 • Clean up consts for keyboard trap
 • First cut at image with larger tooltip on hover
 • Load jQuery UI CSS before our CSS so we can override it
 • Add named function gtt_keydown to prevent duplicate listeners
 • Bump === to work around Google problem
 • Remove Gmail load wait timer now that GtT button is more robust on no-data
 • Move keyboard trap to bindEvents
 • Dirty centering when no data so popup will move as appropriate
 • Fix error with "bottom" should be "below"
 • Refactor upload code to pull it all into model work, add model.Uploader class

chrome/model.js
var GmailToTrello = GmailToTrello || {};

GmailToTrello.Model = function(parent) {
    this.trello = {
        apiKey: 'c50413b23ee49ca49a5c75ccf32d0459',
        user: null,
        boards: null
    };
    this.parent = parent;
    this.settings = {};
    this.isInitialized = false;
    this.event = new EventTarget();
    this.newCard = null;
};

GmailToTrello.Model.prototype.init = function() {
    var self = this;

    this.isInitialized = true;

    // init Trello
    this.initTrello();

};

GmailToTrello.Model.prototype.initTrello = function() {
    // gtt_log("Model:initTrello");

    var self = this;

    self.trello.user = null;
    self.trello.boards = null;

    Trello.setKey(self.trello.apiKey);
    self.checkTrelloAuthorized();
};

GmailToTrello.Model.prototype.checkTrelloAuthorized = function() {
    // gtt_log("checkTrelloAuthorized");

    var self = this;

    // Assures there's a token or not:
    Trello.authorize({
        interactive: false,
        success: function(data) {
            self.event.fire('onAuthorized');
            self.loadTrelloData();
        },
        error: function(data) {
            if (!Trello.authorized()) { // Assure token is invalid
                self.event.fire('onBeforeAuthorize');
                Trello.authorize({
                    type: 'popup',
                    name: "Gmail-to-Trello",
                    interactive: true,
                    persit: true,
                    scope: {read: true, write: true},
                    expiration: 'never',
                    success: function(data) {
                        gtt_log('checkTrelloAuthorized: Trello authorization successful');
                        // gtt_log(data);
                        self.event.fire('onAuthorized');
                        self.loadTrelloData();
                    },
                    error: function(data) {
                        self.event.fire('onAuthorizeFail');
                    }
                });
            } else {
                gtt_log('Model:checkTrelloAuthorized: failed');
                // We have a valid token, so...how did we get here?
                // self.event.fire('onAuthorized');
                // self.loadTrelloData();
                // gtt_log(Trello);
                // gtt_log(Trello.token());
            }
        }
    });
};

GmailToTrello.Model.prototype.deauthorizeTrello = function() {
    gtt_log("deauthorizeTrello");

    Trello.deauthorize();
    this.isInitialized = false;
};

GmailToTrello.Model.prototype.makeAvatarUrl = function(avatarHash) {
    var retn = '';
    if (avatarHash && avatarHash.length > 0) {
        retn = 'https://trello-avatars.s3.amazonaws.com/' + avatarHash + '/30.png';
    }
    return retn;
}

GmailToTrello.Model.prototype.loadTrelloData = function() {
    // gtt_log('loadTrelloData');

    this.event.fire('onBeforeLoadTrello');
    this.trello.user = null;


    var self = this;

    // get user's info
    // gtt_log('loadTrelloData: User info');
    Trello.get('members/me', {}, function(data) {
        if (!data || !data.hasOwnProperty('id')) {
            return false;
        }

        self.trello.user = data;

        // gtt_log('loadTrelloData: User boards');
        self.trello.boards = null;
        Trello.get('members/me/boards', {
                'organization': 'true',
                'organization_fields': 'displayName',
                'filter': 'open',
                'fields': 'name' /* "name,closed" */
            }, function(data) {
                var validData = Array();
                for (var i = 0; i < data.length; i++) {
                    // if (data[i].idOrganization === null)
                    //   data[i].idOrganization = '-1';

                    // Only accept opening boards
                    if (i==0) {
                        // gtt_log(JSON.stringify(data[i]));
                    }
                    if (data[i].closed != true) {
                        validData.push(data[i]);
                    }
                }
                // gtt_log('loadTrelloData: Boards data');
                // gtt_log(JSON.stringify(data));
                // gtt_log(JSON.stringify(validData));
                self.trello.boards = validData;
                self.checkTrelloDataReady();
            }, function failure(data) {
            self.event.fire('onAPIFailure', {data:data});
            }
        );
        self.checkTrelloDataReady();
    }, function failure(data) {
        self.event.fire('onAPIFailure', {data:data});
    });
};

GmailToTrello.Model.prototype.checkTrelloDataReady = function() {
    if (this.trello.user !== null &&
        this.trello.boards !== null) {
        // yeah! the data is ready
        //gtt_log('checkTrelloDataReady: YES');
        //gtt_log(this);
        this.event.fire('onTrelloDataReady');

    }
    //else gtt_log('checkTrelloDataReady: NO');
};


GmailToTrello.Model.prototype.loadTrelloLists = function(boardId) {
    // gtt_log('loadTrelloLists');

    var self = this;
    this.trello.lists = null;

    Trello.get('boards/' + boardId, {lists: "open", list_fields: "name"}, function(data) {
        self.trello.lists = data.lists;
        // gtt_log('loadTrelloLists: lists:' + JSON.stringify(self.trello.lists));
        self.event.fire('onLoadTrelloListSuccess');
    }, function failure(data) {
            self.event.fire('onAPIFailure', {data:data});
    });
};

GmailToTrello.Model.prototype.loadTrelloCards = function(listId) {
    // gtt_log('loadTrelloCards');

    var self = this;
    this.trello.cards = null;

    Trello.get('lists/' + listId + '/cards', {fields: "name,pos,idMembers,idLabels"}, function(data) {
        self.trello.cards = data;
        // gtt_log('loadTrelloCards: cards:' + JSON.stringify(self.trello.cards));
        self.event.fire('onLoadTrelloCardsSuccess');
    }, function failure(data) {
            self.event.fire('onAPIFailure', {data:data});
    });
};

GmailToTrello.Model.prototype.loadTrelloLabels = function(boardId) {
    // gtt_log('loadTrelloLabels');

    var self = this;
    this.trello.labels = null;

    Trello.get('boards/' + boardId + '/labels', {fields: "color,name"}, function(data) {
        self.trello.labels = data;
        // gtt_log('loadTrelloLabels: labels:' + JSON.stringify(self.trello.labels));
        self.event.fire('onLoadTrelloLabelsSuccess');
    }, function failure(data) {
        self.event.fire('onAPIFailure', {data:data});
    });
};

GmailToTrello.Model.prototype.loadTrelloMembers = function(boardId) {
    // gtt_log('loadTrelloMembers');

    var self = this;
    this.trello.members = null;

    Trello.get('boards/' + boardId + '/members', {fields: "fullName,username,initials,avatarHash"}, function(data) {
        var me = self.trello.user;
        // Remove this user from the members list:
        self.trello.members = $.map(data, function (item, iter) {
            return (item.id !== me.id ? item : null);
        });
        // And shove this user in the first position:
        self.trello.members.unshift({
            'id': me.id,
            'username': me.username,
            'initials': me.initials,
            'avatarHash': me.avatarHash,
            'fullName': me.fullName
        });

        // gtt_log('loadTrelloMembers: members:' + JSON.stringify(self.trello.members));

        self.event.fire('onLoadTrelloMembersSuccess');
    }, function failure(data) {
        self.event.fire('onAPIFailure', {data:data});
    });
};

GmailToTrello.Model.prototype.Uploader = function(args) {
    if (!args || !args.hasOwnProperty('parent')) {
        return;
    }

    this.parent = args.parent;

    this.data = [];
    
    this.cardId = args.cardId || '';

    this.pos = this.translatePosition({'position': args.position || '', 'cardPos': args.cardPos || ''});

    if (this.pos !== 'at') {
        this.data.push({'property': 'cards'}); // Seed array for new card
    } 
};

GmailToTrello.Model.prototype.Uploader.prototype = {
    'attachments': 'attachments',

    'exclude': function(list, exclude) {
        let list_new = [];
        $.each(list.split(','), function(iter, item) {
            if (exclude.indexOf(item) === -1) {
                list_new.push(item);
            }
        });
        return list_new.join(',');
    },
    
    'add': function(args) {
        if (this.parent.parent.validHash(args)) {
            if (this.pos !== 'at' && args.property !== this.attachments) { // It's a new card so add to the existing hash:
                this.data[0][args.property] = args.value;
            } else {
                const cardId_k = (this.pos === 'at' ? this.cardId : '%cardId%'); // Won't know until we store the initial card
                args.property = 'cards/' + cardId_k + '/' + args.property;
                this.data.push(args);
            }
        }
        return this;
    },

    'translatePosition': function(args) {
        let pos = 'bottom';
        
        const position_k = args.position || 'below';
        const cardPos_k = parseInt(args.cardPos || 0, 10);

        switch (position_k) {
            case 'below':
                if (cardPos_k) {
                    pos = cardPos_k + 1;
                } else {
                    // pos = 'bottom';
                }
                break;
            case 'to':
                if (!this.cardId || this.cardId.length < 1 || this.cardId === '-1') {
                    pos = 'top';
                } else {
                    pos = 'at';
                }
                break;
        default:
            gtt_log('submit: ERROR: Got unknown case: ' + position_k || '<empty position>');
        }

        return pos;
    },
    
    'process_response': function(data_in) {
        const dl_k = this.parent.parent.deep_link; // Pointer to function for expedience

        const url_k = dl_k(data_in, ['url']);
        const id_k = dl_k(data_in, ['id']);
        const card_k = dl_k(data_in, ['data', 'card']);

        let shortLink = dl_k(card_k, ['shortLink']);
        if (shortLink && shortLink.length > 0) {
            shortLink = 'https://trello.com/c/' + shortLink;
        }
        
        const add_id_k = dl_k(card_k, ['id']);
        const add_title_k = dl_k(card_k, ['name']);

        const new_url_k = shortLink || url_k || '';
        const new_id_k = add_id_k || id_k || '';

        if (new_url_k && this.parent.newCard && !this.parent.newCard.url) {
            this.parent.newCard.url = new_url_k;
        }
        if (new_id_k && this.parent.newCard && !this.parent.newCard.id) {
            this.parent.newCard.id = new_id_k;
            this.cardId = new_id_k;
        }
        if (add_title_k && add_title_k.length > 0) {
            this.parent.newCard.title = add_title_k;
        }
    },

    'attach': function(method, property, upload1, success, failure) {
        if (!property || property.length < 6 || !upload1 || !upload1.value || upload1.value.length < 6) return;
        
        const UPLOAD_ATTACH = 'gtt:upload_attach';
        const UPLOAD_ATTACH_RESULTS = 'gtt:upload_attach_results';
        const trello_url_k = 'https://api.trello.com/1/';
        const param_k = upload1.value;

        // NOTE (Ace, 2020-02-15): We have a funny problem with embedded images so breaking this up:
        // Was: const filename_k = (param_k.split('/').pop().split('#')[0].split('?')[0]) || upload1.name || param_k || 'unknown_filename'; // Removes # or ? after filename
        // Remove # or ? afer filename. Could do this as a regex, but this is a bit faster and more resiliant:
        const slash_split_k = param_k.split('/') // First split by directory slashes
        const end_slash_split_k = slash_split_k[slash_split_k.length - 1] // Take last slash section
        const hash_split_k = end_slash_split_k.split('#') // Split by hash so we can split it off
        const begin_hash_split_k = hash_split_k[0] // Take first hash
        const question_split_k = begin_hash_split_k.split('?') // Now split by question mark to remove variables
        const begin_question_split_k = question_split_k[0] // Take content ahead of question mark
        const filename_k = begin_question_split_k || upload1.name || param_k || 'unknown_filename' // Use found string or reasonable fallbacks

        let callback = function(args) {
            if (args.hasOwnProperty(UPLOAD_ATTACH_RESULTS) && args[UPLOAD_ATTACH_RESULTS] === 'success') {
                success(args);
            } else {
                failure(args);
            }
        }

        let dict = {}
        dict[UPLOAD_ATTACH] = {
            'url_asset': upload1.value,
            'filename': filename_k,
            'trello_key': Trello.key(),
            'trello_token': Trello.token(),
            'url_upload': trello_url_k + property
        }
        
        chrome.runtime.sendMessage(dict, callback);
    },

    'upload': function() {
        let upload1 = this.data.shift();
        if (!upload1) {
            this.event.fire('onCardSubmitComplete');
        } else {
            let generateKeysAndValues = function(object) {
                let keysAndValues = [];
                $.each(object, function(key, value) {
                    keysAndValues.push(key + '=' + (value || '') + ' (' + (value || '').toString().length + ')');
                });
                return keysAndValues.sort().join(' ');
            };

            const dict_k = {'cardId': this.cardId || ''};

            let method = upload1.method || 'post';
            let property = this.parent.parent.replacer(upload1.property, dict_k);
            delete upload1.method;
            delete upload1.property;

            const fn_k = property.endsWith(this.attachments) ? this.attach : Trello.rest;

            let self = this;
            fn_k(method, property, upload1, function success(data) {
                $.extend(data, {'method': method + ' ' + property, 'keys': generateKeysAndValues(upload1)});
                self.process_response(data);
                if (self.data && self.data.length > 0) {
                    self.upload();
                } else {
                    self.parent.event.fire('onCardSubmitComplete', {data: data});
                }
            }, function failure(data) {
                $.extend(data, {'method': method + ' ' + property, 'keys': generateKeysAndValues(upload1)});
                self.parent.event.fire('onAPIFailure', {data: data});
            });
        }
    }

};

GmailToTrello.Model.prototype.submit = function() {
    let self = this;
    if (self.newCard === null) {
        gtt_log('submit: data is empty');
        return false;
    }

    self.parent.saveSettings();

    var data = self.newCard;
    
    var text = data.title || '';
    if (text.length > 0) {
        if (data.markdown) {
            text = '**' + text + '**\n\n';
        }
    }
    text += data.description;

    text = self.parent.truncate(text, self.parent.popupView.MAX_BODY_SIZE, '...');

    var desc = self.parent.truncate(data.description, self.parent.popupView.MAX_BODY_SIZE, '...');

    var due_text = '';

    if (data.due_Date && data.due_Date.length > 1) { // Will 400 if not valid date:
        /* Workaround for quirk in Date object,
         * See: http://stackoverflow.com/questions/28234572/html5-datetime-local-chrome-how-to-input-datetime-in-current-time-zone
         * Was: dueDate.replace('T', ' ').replace('-','/')
         */
        let due = data.due_Date.replace('-', '/');

        if (data.due_Time && data.due_Time.length > 1) {
            due += ' ' + data.due_Time;
        } else {
            due += ' 00:00'; // Must provide time
        }
        due_text = (new Date(due)).toISOString();
        /* (NOTE (Ace, 27-Feb-2017): When we used datetime-local object, this was:
        trelloPostableData.due = (new Date(data.dueDate.replace('T', ' ').replace('-','/'))).toISOString();
        */
    }

    let uploader = new this.Uploader({
        'parent': self, 
        'cardId': data.cardId,
        'position': data.position,
        'cardPos': data.cardPos
    });

    const pos_k = uploader.pos;

    if (pos_k === 'at') {
        uploader.add({'property': 'actions/comments', 'text': text});
    } else {
        uploader
            .add({'property': 'pos', 'value': pos_k})
            .add({'property': 'name', 'value': data.title})
            .add({'property': 'desc', 'value': desc})
            .add({'property': 'idList', 'value': data.listId});
    }

    uploader
        .add({'property': 'idMembers', 'value': uploader.exclude(data.membersId, data.cardMembers)})
        .add({'property': 'idLabels', 'value': uploader.exclude(data.labelsId, data.cardLabels)})
        .add({'property': 'due', 'value': due_text, 'method': 'put'})

    let imagesAndAttachments = (data.images || []).concat(data.attachments || []);

    $.each(imagesAndAttachments, function(iter, item) {
        if (item.hasOwnProperty('checked') && item.checked && item.url && item.url.length > 5) {
            uploader.add({'property': uploader.attachments, 'value': item.url, 'name': item.name});
        }
    });

  uploader.upload();
};

// End, model.js